### ISSUE TITLE
Invalid Function Signatures for pybind modules

### ISSUE THREAD

**Issue by AngelJA** [pytorch/pytorch#88958]
### 🐛 Describe the bug

Current use of `PYBIND11_TYPE_CASTER` produces function signatures that are not valid python.

For example, adding `print(cpp_extension.sigmoid_add.__doc__)` somewhere in `test/test_cpp_extensions_aot.py`, then running with `python run_test.py -i test_cpp_extensions_aot_ninja` shows
```
sigmoid_add(arg0: at::Tensor, arg1: at::Tensor) -> at::Tensor

sigmoid(x) + sigmoid(y)
```

This causes errors when used with tools like https://github.com/sizmailov/pybind11-stubgen.

My proposed solution would be changing lines in `torch/csrc/utils/pybind.h` like 
`PYBIND11_TYPE_CASTER(at::Tensor, _("at::Tensor"));`
to
`PYBIND11_TYPE_CASTER(at::Tensor, _("torch.Tensor"));`

Is this a change we could make? Could this mess something up that I'm not aware of?

### Versions

Collecting environment information...
PyTorch version: 1.14.0a0+gitba4d5aa
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.31

Python version: 3.9.13 (main, Aug 25 2022, 23:26:10)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-4.4.0-19041-Microsoft-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.5
[pip3] numpydoc==1.4.0
[pip3] torch==1.14.0a0+gitba4d5aa
[pip3] torch-test-cpp-extension==0.0.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-include               2022.1.0           h06a4308_224
[conda] mkl-service               2.4.0            py39h7f8727e_0
[conda] mkl_fft                   1.3.1            py39hd3c417c_0
[conda] mkl_random                1.2.2            py39h51133e4_0
[conda] numpy                     1.21.5           py39h6c91a56_3
[conda] numpy-base                1.21.5           py39ha15fc14_3
[conda] numpydoc                  1.4.0            py39h06a4308_0
[conda] torch                     1.14.0a0+gitba4d5aa           dev_0    <develop>
[conda] torch-test-cpp-extension  0.0.0                     dev_0    <develop>

cc @ezyang @bhosmer @smessmer @ljk53 @bdhirsh

**Comment by ezyang** [comment]
Yes, send us the patch

**Comment by AngelJA** [comment]
Thanks, just opened a PR! Any idea how we could add test coverage, or if it's needed?

Tried figuring out the right types through exploration. Although I wasn't able to create bindings for `SymInt` and `SymFloat` because of a linker error.
`m.def("f", []() { return at::SymInt(); });`
```
/home/aavila/pytorch/torch/include/pybind11/pybind11.h:249: undefined reference to `pybind11::detail::type_caster<c10::SymInt, void>::cast(c10::SymInt, pybind11::return_value_policy, pybind11::handle)'
/home/aavila/anaconda3/compiler_compat/ld: build/lib.linux-x86_64-cpython-39/torch_test_cpp_extension/cpp.cpython-39-x86_64-linux-gnu.so: hidden symbol `_ZN8pybind116detail11type_casterIN3c106SymIntEvE4castES3_NS_19return_value_policyENS_6handleE' isn't defined
/home/aavila/anaconda3/compiler_compat/ld: final link failed: bad value
```

**Comment by ezyang** [comment]
If you want to skip tests I don't mind. Otherwise something simple like checking each of the sigs are valid Python

**Comment by ezyang** [comment]
Not sure what's up with the linker error

**Comment by AngelJA** [comment]
Since making this issue I've realized a great way to demonstrate the issue & fix is using `mypy.stubgen`.

After everything is built (including `test/cpp_extensions`), running this command:
```
python -c "import torch; from mypy.stubgen import main; main(['-m', 'torch_test_cpp_extension.cpp'])"
```
Used to produce this stub file:
```python
from typing import Any

class MatrixMultiplier:
    def __init__(self, arg0: int, arg1: int) -> None: ...
    def forward(self, *args, **kwargs) -> Any: ...
    def get(self, *args, **kwargs) -> Any: ...

def function_taking_optional(arg0) -> bool: ...
def sigmoid_add(*args, **kwargs) -> Any: ...
```
After my PR it produces this stub file:
```python
from typing import Optional

import torch

class MatrixMultiplier:
    def __init__(self, arg0: int, arg1: int) -> None: ...
    def forward(self, arg0: torch.Tensor) -> torch.Tensor: ...
    def get(self) -> torch.Tensor: ...

def function_taking_optional(arg0: Optional[torch.Tensor]) -> bool: ...
def sigmoid_add(arg0: torch.Tensor, arg1: torch.Tensor) -> torch.Tensor: ...
```
Definitely an improvement! But, the reason for this post is I think not all type_casters in `pybind.h` will play nicely like this, there's still room for improvement. Also pondering if there's a way to make a test out of this and ensure good coverage.

**Comment by ezyang** [comment]
Seems reasonable. test_typing is a good spot

**Comment by AngelJA** [comment]
@ezyang Sorry for the delay! Finally got around to opening a PR with some fixes and tests: #89897
